src/base/http/requestparser.cpp

```diff
@@ -180,11 +180,14 @@ bool RequestParser::parseRequestLine(const QString &line)
     m_request.method = match.captured(1);
 
     // Request Target
-    const QByteArray decodedUrl {QByteArray::fromPercentEncoding(match.captured(2).toLatin1())};
-    const int sepPos = decodedUrl.indexOf('?');
-    m_request.path = QString::fromUtf8(decodedUrl.constData(), (sepPos == -1 ? decodedUrl.size() : sepPos));
+    // URL components should be separated before percent-decoding
+    // [rfc3986] 2.4 When to Encode or Decode
+    const QByteArray url {match.captured(2).toLatin1()};
+    const int sepPos = url.indexOf('?');
+    const QByteArray pathComponent = QByteArray::fromRawData(url.constData(), (sepPos == -1 ? url.size() : sepPos));
+    m_request.path = QString::fromUtf8(QByteArray::fromPercentEncoding(pathComponent));
     if (sepPos >= 0)
-        m_request.query = decodedUrl.mid(sepPos + 1);
+        m_request.query = url.mid(sepPos + 1);
```

Comment:
No, of course. Otherwise it will not achieve its goals.
Query string may contain encoded special characters, so the application should parse it before decoding, otherwise it may not do it correctly. E.g:
```
paramName=foo%26bar
```
There is only one parameter "paramName" with value "foo&bar". But if you decode it first then it will be parsed as two parameters ("paramName" with value "foo" and unnamed param with value "bar").

Reply:
OK, I see.